urlcdd0c1ab49_emiln_JHouseFireExplorer_tgz-pJ8-dk_brics_automaton_DatatypesJ8/src/dk/brics/automaton/RunAutomaton.java

```diff
@@ -0,0 +1,296 @@
+
+	/** 
+	 * Sets alphabet table for optimal run performance. 
+	 */
+	final void setAlphabet() {
+		classmap = new int[Character.MAX_VALUE - Character.MIN_VALUE + 1];
+		int i = 0;
+		for (int j = 0; j <= Character.MAX_VALUE - Character.MIN_VALUE; j++) {
+			if (i + 1 < points.length && j == points[i + 1])
+				i++;
+			classmap[j] = i;
+		}
+	}
+
+	/** 
+	 * Returns a string representation of this automaton. 
+	 */
+	@Override
+	public String toString() {
+		StringBuilder b = new StringBuilder();
+		b.append("initial state: ").append(initial).append("\n");
+		for (int i = 0; i < size; i++) {
+			b.append("state " + i);
+			if (accept[i])
+				b.append(" [accept]:\n");
+			else
+				b.append(" [reject]:\n");
+			for (int j = 0; j < points.length; j++) {
+				int k = transitions[i * points.length + j];
+				if (k != -1) {
+					char min = points[j];
+					char max;
+					if (j + 1 < points.length)
+						max = (char)(points[j + 1] - 1);
+					else
+						max = Character.MAX_VALUE;
+					b.append(" ");
+					Transition.appendCharString(min, b);
+					if (min != max) {
+						b.append("-");
+						Transition.appendCharString(max, b);
+					}
+					b.append(" -> ").append(k).append("\n");
+				}
+			}
+		}
+		return b.toString();
+	}
+
+	/** 
+	 * Returns number of states in automaton. 
+	 */
+	public int getSize() {
+		return size;
+	}
+
+	/** 
+	 * Returns acceptance status for given state. 
+	 */
+	public boolean isAccept(int state) {
+		return accept[state];
+	}
+
+	/** 
+	 * Returns initial state. 
+	 */
+	public int getInitialState() {
+		return initial;
+	}
+
+	/**
+	 * Returns array of character class interval start points. The array should
+	 * not be modified by the caller.
+	 */
+	public char[] getCharIntervals() {
+		return points.clone();
+	}
+
+	/** 
+	 * Gets character class of given char. 
+	 */
+	int getCharClass(char c) {
+		return SpecialOperations.findIndex(c, points);
+	}
+
+	@SuppressWarnings("unused")
+	private RunAutomaton() {}
+
+	/**
+	 * Constructs a new <code>RunAutomaton</code> from a deterministic
+	 * <code>Automaton</code>. Same as <code>RunAutomaton(a, true)</code>.
+	 * @param a an automaton
+	 */
+	public RunAutomaton(Automaton a) {
+		this(a, true);
+	}
+
+	/**
+	 * Retrieves a serialized <code>RunAutomaton</code> located by a URL.
+	 * @param url URL of serialized automaton
+	 * @exception IOException if input/output related exception occurs
+	 * @exception OptionalDataException if the data is not a serialized object
+	 * @exception InvalidClassException if the class serial number does not match
+	 * @exception ClassCastException if the data is not a serialized <code>RunAutomaton</code>
+	 * @exception ClassNotFoundException if the class of the serialized object cannot be found
+	 */
+	public static RunAutomaton load(URL url) throws IOException, OptionalDataException, ClassCastException, 
+													ClassNotFoundException, InvalidClassException {
+		return load(url.openStream());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *openStream*. Currently, there are execution paths that do not contain closure statements, for example, when RunAutomaton.load throws an exception. Ensure the object returned by *openStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).